tempfile.1
.\" -*- nroff -*-
.TH TEMPFILE 1 "30 May 2011" "Debian"
.SH NAME
tempfile \- create a temporary file in a safe manner
.SH SYNOPSIS
.B tempfile
[\-d DIR] [\-p STRING] [\-s STRING] [\-m MODE] [\-n FILE] [\-\-directory=DIR]
[\-\-prefix=STRING] [\-\-suffix=STRING] [\-\-mode=MODE] [\-\-name=FILE] [\-\-help] [\-\-version]
.SH DESCRIPTION
.PP
.B tempfile
creates a temporary file in a safe manner.  It uses
.BR tempnam (3)
to choose the name and opens it with O_RDWR | O_CREAT | O_EXCL.  The filename
is printed on standard output. See
.BR tempnam (3)
for the actual steps involved in directory selection.
.PP
The directory in which to create the file might be searched for in this
order (but refer to
.BR tempnam (3)
for authoritative answers):
.TP 3
a)
In case the environment variable
.B TMPDIR
exists and
contains the name of an appropriate directory, that is used.
.TP
b)
Otherwise, if the
.I \-\-directory
argument is specified and appropriate, it is used.
.TP
c)
Otherwise,
.I P_tmpdir
(as defined in
.IR <stdio.h> )
is used when appropriate.
.TP
d)
Finally an implementation-defined directory
.IR (/tmp)
may be used.
.SH OPTIONS
.TP
.BI "-d, --directory " DIR
Place the file in DIR.
.TP
.BI "-m, --mode " MODE
Open the file with MODE instead of 0600.
.TP
.BI "-n, --name " FILE
Use FILE for the name instead of
.BR tempnam (3) .
The options -d, -p, and -s are ignored if this option is given.
.TP
.BI "-p, --prefix " STRING
Use up to five letters of STRING to generate the name.
.TP
.BI "-s, --suffix " STRING
Generate the file with STRING as the suffix.
.TP
.B "--help"
Print a usage message on standard output and exit successfully.
.TP
.B "--version"
Print version information on standard output and exit successfully.
.SH RETURN VALUES
An exit status of 0 means the temporary file was created successfully.
Any other exit status indicates an error.
.SH BUGS
Exclusive creation is not guaranteed when creating files on NFS
partitions.  tempfile is deprecated; you should use mktemp(1)
instead.
.SH EXAMPLE

.RS
.nf
#!/bin/sh
#[...]
t=$(tempfile) || exit
trap "rm -f -- '$t'" EXIT
#[...]
rm -f -- "$t"
trap - EXIT
exit
.fi
.SH "SEE ALSO"
.BR tempnam (3),
.BR mktemp (1)


debian/changelog
debianutils (4.0.1) unstable; urgency=low

  * tempfile.1: state in the man page that tempfile is
    deprecated.  closes: #627564.
  * ischroot.1: correct root/non-root logic claims.
    closes: #628465.

 -- Clint Adams <clint@debian.org>  Mon, 30 May 2011 10:54:03 -0400

debianutils (4) unstable; urgency=low

  * ischroot utility by Aurelien Jarno to detect chrootedness.
    closes: #627205.

 -- Clint Adams <clint@debian.org>  Wed, 18 May 2011 15:04:15 -0400

debianutils (3.4.6) unstable; urgency=low

  * Add Spanish man page translations from Omar
    Campagne.  closes: #619016.
  * Updated German man page translations from Helge
    Kreutzmann.  closes: #626891.

 -- Clint Adams <clint@debian.org>  Tue, 17 May 2011 23:29:42 -0400

debianutils (3.4.5) unstable; urgency=low

  * Vcs-Git update.
  * add-shell(8): grammar fix from Helge Kreutzmann.  closes: #626409.
  * German man page translations from Helge Kreutzmann.  closes: #626397.
  * Bump to Standards-Version 3.9.2.

 -- Clint Adams <clint@debian.org>  Thu, 12 May 2011 22:45:15 -0400

debianutils (3.4.4) unstable; urgency=low

  * French man page translation update from David Prévot.
    closes: #603908.
  * Patch from Jari Aalto to reorder tempfile options alphabetically.
    closes: #600542.

 -- Clint Adams <clint@debian.org>  Thu, 10 Feb 2011 17:24:34 -0500

debianutils (3.4.3) unstable; urgency=low

  * Patch from Ben Hutchings to have installkernel run hook scripts.
    closes: #607411.

 -- Clint Adams <clint@gnu.org>  Tue, 18 Jan 2011 15:29:35 -0500

debianutils (3.4.2) unstable; urgency=low

  * Add Slovene man page translations from Peter Kragelj.  closes:
    #598531.
  * French man page translation update from Christian Perrier.  closes:
    #597486.
  * run-parts.8: be more precise about ASCII requirement in naming.
    closes: #542029.

 -- Clint Adams <clint@gnu.org>  Sun, 14 Nov 2010 10:47:12 -0500

debianutils (3.4.1) unstable; urgency=low

  * po4a/po/fr.po: French man page translation update from
    Christian Perrier.  closes: #597430.
  * Bump to Standards-Version 3.9.1.

 -- Clint Adams <schizo@debian.org>  Sun, 19 Sep 2010 12:58:38 -0400

debianutils (3.4) unstable; urgency=low

  * Apply patch from Ariel to add strength options for compression and
    reduce xz default compression level.  closes: #587766.
  * Bump Standards-Version to 3.9.0.

 -- Clint Adams <schizo@debian.org>  Thu, 15 Jul 2010 22:02:27 -0400

debianutils (3.3) unstable; urgency=low

  * savelog.8: add SEE ALSO for logrotate(8).  closes: #487171.
  * savelog: add -J option for xz compression.  closes: #587520.

 -- Clint Adams <schizo@debian.org>  Wed, 30 Jun 2010 10:38:10 -0400

debianutils (3.2.3) unstable; urgency=low

  * Switch to 3.0 (native) source format.
  * French man page translation updates from Alexandre Hoïde.
    closes: #579652.
  * Bump to Standards-Version 3.8.4.

 -- Clint Adams <schizo@debian.org>  Thu, 29 Apr 2010 11:12:26 -0400

debianutils (3.2.2) unstable; urgency=low

  * savelog: patch from Patrick Coleman to fix breakage when target file
    does not exist.  closes: #556227.
  * Bump to Standards-Version 3.8.3.

 -- Clint Adams <schizo@debian.org>  Sat, 14 Nov 2009 13:16:01 -0500

debianutils (3.2.1) unstable; urgency=low

  * run-parts.8: conform man page regex with the actual regex used by
    run-parts.  closes: #536395.

 -- Clint Adams <schizo@debian.org>  Sun, 09 Aug 2009 15:18:52 -0400

debianutils (3.2) unstable; urgency=low

  * run-parts.c: print command arguments when --verbose and --arg are
    used.  closes: #535621.
  * Bump to Standards-Version 3.8.2.

 -- Clint Adams <schizo@debian.org>  Sun, 05 Jul 2009 15:03:00 -0400

debianutils (3.1.3) unstable; urgency=low

  * Remove any mention of readlink from debian/copyright.  closes:
    #527138.

 -- Clint Adams <schizo@debian.org>  Tue, 05 May 2009 21:08:31 -0400

debianutils (3.1.2) unstable; urgency=low

  * Patch from Zack Weinberg to fix tempfile -n crash.  closes: #526981.

 -- Clint Adams <schizo@debian.org>  Mon, 04 May 2009 18:56:14 -0400

debianutils (3.1.1) unstable; urgency=medium

  [ Manoj Srivastava ]
  * Fix a double free bug in tempfile. Closes: #526606, #526607, #526610

 -- Manoj Srivastava <srivasta@debian.org>  Sat, 02 May 2009 03:45:30 -0500

debianutils (3.1) unstable; urgency=medium

  [ Manoj Srivastava ]
  * which.1: clarify that which does not follow symlinks.  closes: #268624.
  * run-parts.c: permission and sanity checks run on link targets, not
    symlinks.  closes: #294789.
  * tempfile.c: tempfile with a suffix no longer needs hardlinks.
    closes: #312270.
  * savelog.8: disambiguate version numbering in manual.  closes: #330289.
  * tempfile.1: update manual page with up-to-date tempnam behaviors.
    closes: #338490.

  [ Clint Adams ]
  * savelog.8, tempfile.1: remove spurious trailing whitespace
  * savelog.8, tempfile.1, which.1: update dates in manual pages.
  * po4a/po/fr.po: tweak French translations enough to reach 80%.

 -- Clint Adams <schizo@debian.org>  Fri, 01 May 2009 18:01:28 -0400

debianutils (3.0.2) unstable; urgency=low

  * installkernel: fix symlinking of kernel images.  closes: #526493.

 -- Clint Adams <schizo@debian.org>  Fri, 01 May 2009 11:08:13 -0400

debianutils (3.0.1) unstable; urgency=low

  * Remove French sensible-editor man page as well.  closes: #521543.

 -- Clint Adams <schizo@debian.org>  Sun, 29 Mar 2009 17:55:21 -0400

debianutils (3.0) unstable; urgency=low

  * Stop shipping sensible-browser, sensible-editor, sensible-pager, and
    accompanying man pages, and declare a dependency on sensible-utils,
    to which they have moved.

 -- Clint Adams <schizo@debian.org>  Thu, 26 Mar 2009 21:25:08 -0400

debianutils (2.32.3) unstable; urgency=medium

  * installkernel: remove stray right bracket, thanks to Alban Browaeys.
    closes: #521286.

 -- Clint Adams <schizo@debian.org>  Thu, 26 Mar 2009 09:53:46 -0400

debianutils (2.32.2) unstable; urgency=low


tempfile.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif /* HAVE_GETOPT_H */
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>

char *progname;

void usage(int);
void syserror(const char *);
int parsemode(const char *, mode_t *);

void
usage (int status)
{
  if (status)
    fprintf(stderr, "Try `%s --help' for more information.\n", progname);
  else
    printf("Usage: %s [OPTION]\n\n"
"Create a temporary file in a safe manner.\n\n"
"-d, --directory=DIR  place temporary file in DIR\n"
"-m, --mode=MODE      open with MODE instead of 0600\n"
"-n, --name=FILE      use FILE instead of tempnam(3)\n"
"-p, --prefix=STRING  set temporary file's prefix to STRING\n"
"-s, --suffix=STRING  set temporary file's suffix to STRING\n"
"    --help           display this help and exit\n"
"    --version        output version information and exit\n", progname);
  exit(status);
}


void
syserror (const char *fx)
{
  perror(fx);
  exit(1);
}


int
parsemode (const char *in, mode_t *out)
{
  char *endptr;
  long int mode;
  mode = strtol(in, &endptr, 8);
  if (*endptr || mode<0 || mode>07777)
    return 1;
  *out = (mode_t) mode;
  return 0;
}


int
main (int argc, char **argv)
{
  char *name=0, *dir=0, *pfx=0, *sfx=0, *filename=0;
  mode_t mode = 0600;
  int fd, optc;
  struct option long_options[] = {
    {"prefix", required_argument, 0, 'p'},
    {"suffix", required_argument, 0, 's'},
    {"directory", required_argument, 0, 'd'},
    {"mode", required_argument, 0, 'm'},
    {"name", required_argument, 0, 'n'},
    {"help", no_argument, 0, 'h'},
    {"version", no_argument, 0, 'v'},
    {0, 0, 0, 0}
  };
  progname = argv[0];

  while ((optc = getopt_long (argc, argv, "p:s:d:m:n:", long_options, 0))
	 != EOF) {
    switch (optc) {
    case 0:
      break;
    case 'p':
      pfx = optarg;
      break;
    case 's':
      sfx = optarg;
      break;
    case 'd':
      dir = optarg;
      break;
    case 'm':
      if (parsemode(optarg, &mode)) {
	fprintf(stderr, "Invalid mode `%s'.  Mode must be octal.\n", optarg);
	usage(1);
      }
      break;
    case 'n':
      /* strdup because it is freed later on */
      if((name = strdup(optarg)) == NULL)
        syserror("strdup");
      break;
    case 'h':
      usage(0);
    case 'v':
      puts("tempfile " PACKAGE_VERSION);
      exit(0);
    default:
      usage(1);
    }
  }

  if (name) {
    if ((fd = open(name, O_RDWR | O_CREAT | O_EXCL, mode)) < 0)
      syserror("open");
    filename = name;
  }
  else {
    for (;;) {
      if (!(name = tempnam(dir, pfx)))
	syserror("tempnam");
      if(sfx) {
        char *namesfx;
        if (!(namesfx = (char *)malloc(strlen(name) + strlen(sfx) + 1)))
          syserror("malloc");
        strcpy(namesfx, name);
        strcat(namesfx, sfx);
        filename = namesfx;
      }
      else
        filename = name;

      if ((fd = open(filename, O_RDWR | O_CREAT | O_EXCL, mode)) < 0) {
	if (errno == EEXIST) {
          if(name != filename)
            free(name);
	  free(filename);
	  continue;
	}
	syserror("open");
      }
      break;
    }
  }
  
  if (close(fd))
    syserror("close");
  puts(filename);
  if(name != filename)
    free(name);
  free(filename);
  exit(0);
}
